tracing-appender/src/non_blocking.rs

```diff
@@ -144,3 +144,124 @@ impl Drop for WorkerGuard {
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use rand::Rng;
+    use std::sync::Mutex;
+    use std::thread;
+    use std::time::Duration;
+
+    struct MockWriter {
+        writer: Arc<Mutex<Vec<String>>>,
+        max_writes_allowed: usize,
+        writes_attempted: usize,
+    }
+
+    impl Default for MockWriter {
+        fn default() -> Self {
+            MockWriter {
+                writer: Arc::new(Mutex::new(Vec::new())),
+                max_writes_allowed: 1,
+                writes_attempted: 0,
+            }
+        }
+    }
+
+    impl std::io::Write for MockWriter {
+        fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+            let buf_len = buf.len();
+            self.writes_attempted += 1;
+            if self.writes_attempted > self.max_writes_allowed {
+                return Err(std::io::Error::from(std::io::ErrorKind::WouldBlock));
+            }
+
+            self.writer
+                .lock()
+                .expect("expected guard")
+                .push(String::from_utf8_lossy(buf).to_string());
+            Ok(buf_len)
+        }
+
+        fn flush(&mut self) -> std::io::Result<()> {
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn logs_dropped_if_lossy() {
+        let mock_writer = MockWriter::default();
+
+        let (mut non_blocking, _guard) = self::NonBlockingBuilder::default()
+            .lossy(true)
+            .buffered_lines_limit(1)
+            .finish(mock_writer);
+
+        let error_count = non_blocking.error_counter();
+
+        non_blocking
+            .write_all("Hello".as_bytes())
+            .expect("Failed to write");
+        assert_eq!(0, error_count.load(Ordering::Relaxed));
+
+        non_blocking
+            .write_all(", World".as_bytes())
+            .expect("Failed to write");
+        assert_eq!(1, error_count.load(Ordering::Relaxed));
+
+        non_blocking.write_all(".".as_bytes()).expect("Failed to write");
+        assert_eq!(2, error_count.load(Ordering::Relaxed));
+    }
+
+    #[test]
+    fn multi_threaded_writes() {
```

Comment:
I am not sure if we need loom for this, but I am a little confused about the intention behind the `thread::sleep` — depending on what that's supposed to be doing, it could be good to replace it with loom to explore interleavings more exhaustively? But, I'm not totally sure that it;s necessary.

Reply:
It looks like this test is failing on [one of the CI test jobs](https://github.com/tokio-rs/tracing/pull/678/checks?check_run_id=593985757#step:4:449) but passing on others. That's pretty concerning. We should make sure this is not racy before moving forward with it.

Reply:
I think we can resolve this comment now.